src/index.ts

```diff
@@ -1,132 +1,231 @@
 
-let codec: WebAssembly.Instance;
+let codec: WebAssembly.Instance
 
 // Init the library with the binary content of a .wasm file.
 // This function must be called once per client prior to use of any other package functions.
-export async function init(code: BufferSource) {
+export async function init(code: BufferSource) : Promise<void> {
     const wasm = await WebAssembly.compile(code);
     codec = await WebAssembly.instantiate(wasm, {})
 }
 
-export function isInitialized() {
-    return (codec !== null)
+// Returns true iff library was initialized with Wasm code
+export function isInitialized() : boolean {
+    return (codec !== undefined)
 }
 
-// Call an svm_codec function with the provided buffer. Returns result buffer.
-export function call(funcName: string, buf) {
-    return (codec!.exports as any)[funcName](buf)
+// User input to the encodeSpawnData function
+export interface Spawn {
+    version: number,
+    template: string,
+    name: string
+    ctor_name: string,
+    calldata: Uint8Array,
 }
 
-// Allocates a svm_codec buffer with the provided byte length
-export function wasmBufferAlloc(length: number) {
-    return (codec!.exports as any).wasm_alloc(length)
+// User input to the encodeCallData function
+export interface Call {
+    version: number,
+    target: string,
+    func_name: string,
+    verifydata: Uint8Array,
+    calldata: Uint8Array,
+}
+
+
+// Encodes the provided spawn app data
+export function encodeSpawn(data: Spawn) : Uint8Array {
+    const buf = newWasmBuffer(data);
+    const result = call("wasm_encode_spawn", buf);
+    const len = wasmBufferLength(result);
+    const slice = wasmBufferDataSlice(result, 0, len);
+
+    let errMessage = "";
+    if (slice[0] !== OK_MARKER) {
+        errMessage = loadWasmBufferError(result);
+    }
+
+    wasmBufferFree(buf);
+    wasmBufferFree(result);
+
+    if (errMessage.length > 0) {
+        throw errMessage;
+    }
+
+    return slice.slice(1);
+}
+
+// Decodes the provided SVM encoded spawn data
+export function decodeSpawn(data: Uint8Array) : JSON {
+    const buf = newWasmBuffer({data: binToString(data)});
+    const result = call("wasm_decode_spawn", buf);
+    const json = loadWasmBufferDataAsJson(result);
+    wasmBufferFree(buf);
+    wasmBufferFree(result);
+    return json;
+}
+
+// Encodes the provided call account data
+export function encodeCall(data:Call) : Uint8Array {
+    const buf = newWasmBuffer(data);
+    const result = call("wasm_encode_call", buf);
+    const len = wasmBufferLength(result);
+    const slice = wasmBufferDataSlice(result, 0, len);
+
+    let errMessage = "";
+    if (slice[0] !== OK_MARKER) {
+        errMessage = loadWasmBufferError(result);
+    }
+
+    wasmBufferFree(buf);
+    wasmBufferFree(result);
+
+    if (errMessage.length > 0) {
+        throw errMessage;
+    }
+
+    return slice.slice(1);
+}
+
+// Decodes the SVM encoded call account data
+export function decodeCall(bytes: Uint8Array): JSON {
```

Comment:
Both `decodeCall`, `decodeSpawn` and `decodeInput` behave pretty much the same. 
The only difference is the function name passed the `call.`

Probably we'd be better with extracting the sharde code into a `decodeMessage` function.

Reply:
I don't think that further refactoring should be part of this PR. It should be done IMO in a follow-up PR.